lightning/src/ln/onchain_utils.rs

```diff
@@ -0,0 +1,892 @@
+
+		Ok(())
+	}
+}
+
+impl Readable for OnchainRequest {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		let aggregation = Readable::read(reader)?;
+		let bump_strategy = Readable::read(reader)?;
+		let feerate_previous = Readable::read(reader)?;
+		let height_timer = Readable::read(reader)?;
+		let absolute_timelock = Readable::read(reader)?;
+		let height_original = Readable::read(reader)?;
+		let content = Readable::read(reader)?;
+
+		Ok(OnchainRequest {
+			aggregation,
+			bump_strategy,
+			feerate_previous,
+			height_timer,
+			absolute_timelock,
+			height_original,
+			content
+		})
+	}
+}
+
+fn subtract_high_prio_fee<F: Deref, L: Deref>(input_amounts: u64, predicted_weight: usize, fee_estimator: &F, logger: &L) -> Option<(u64, u64)>
+	where F::Target: FeeEstimator,
+	      L::Target: Logger,
+{
+	let mut updated_feerate = fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::HighPriority) as u64;
+	let mut fee = updated_feerate * (predicted_weight as u64) / 1000;
+	if input_amounts <= fee {
+		updated_feerate = fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::Normal) as u64;
+		fee = updated_feerate * (predicted_weight as u64) / 1000;
+		if input_amounts <= fee {
+			updated_feerate = fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::Background) as u64;
+			fee = updated_feerate * (predicted_weight as u64) / 1000;
+			if input_amounts <= fee {
+				log_error!(logger, "Failed to generate an on-chain punishment tx as even low priority fee ({} sat) was more than the entire claim balance ({} sat)",
+					fee, input_amounts);
+				None
+			} else {
+				log_warn!(logger, "Used low priority fee for on-chain punishment tx as high priority fee was more than the entire claim balance ({} sat)",
+					input_amounts);
+				Some((fee, updated_feerate))
+			}
+		} else {
+			log_warn!(logger, "Used medium priority fee for on-chain punishment tx as high priority fee was more than the entire claim balance ({} sat)",
+				input_amounts);
+			Some((fee, updated_feerate))
+		}
+	} else {
+		Some((fee, updated_feerate))
+	}
+}
+
+fn feerate_bump<F: Deref, L: Deref>(predicted_weight: usize, input_amounts: u64, previous_feerate: u64, fee_estimator: &F, logger: &L) -> Option<(u64, u64)>
+	where F::Target: FeeEstimator,
+	      L::Target: Logger,
+{
+	// If old feerate inferior to actual one given back by Fee Estimator, use it to compute new fee...
+	let new_fee = if previous_feerate < fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::HighPriority) as u64 {
+		if let Some((new_fee, _)) = subtract_high_prio_fee(input_amounts, predicted_weight, fee_estimator, logger) {
+			new_fee
+		} else {
+			log_trace!(logger, "Can't new-estimation bump new claiming tx, amount {} is too small", input_amounts);
+			return None;
+		}
+	// ...else just increase the previous feerate by 25% (because that's a nice number)
+	} else {
+		let fee = previous_feerate * (predicted_weight as u64) / 750;
```

Comment:
I think you're right :

* (feerate=1000) * (weight=100) / (weight_denominator=1000) = 100
* (feerate=1000) * (weight=100) / (weight_denominator=750) = 133

So good catch! I need to update tests so not marking at resolved yet. If you have a better idea on the % heuristic lmk it's quite arbitrary rn :)

Reply:
well, to be obvious, this has to be high enough so it tracks feerate spikes within the expiration time, especially when there's an attack on the mempool.  if I understand the code correctly, a bump happens on every block, right?

so, the desired bump % is dependent on the expiration time in blocks and the ratio of the spike to original feerate.  one issue is if the original feerate was very low, let's say because the network was quiet.  then perhaps we can get a very high spike ratio. 

for example, with a 100x spike height and 144 block expiration, the bump should be around 3.2%.  we could handle much higher spikes with a 10% increase - up to a `(1.1 ^ 144) = 900000` feerate increase over 144 blocks.

we should have this in a const with some documentation on the const justifying the value based on assumptions on maximum spike height and minimum expiration time.

Reply:
Bump happens accordingly to `OnchainTxHandler::get_height_timer`, if expiration is coming in next 3 blocks, bumps at every block.

What do you qualify as spike height here ? The original height at which the transaction is broadcast for first time ? Ideally you want the fee-bump to be exponential, save on fees if timelock is far away, be ready to burn as funds at stake when the expiration is really near.

Agree on documenting our assumptions.

Reply:
apologies, missed this reply.  I should have said "spike factor" - which I meant as the ratio of the feerate when we start to try to publish and the feerate at the peak of the spike.

I agree that the bumps should be exponential (and they are in your code, right?).

But my assumption about the bump frequency was incorrect.  Looking at `get_height_timer`, if we have 144 blocks, we bump 9 times every 15 blocks, then twice every 3 blocks then another 3 times, for a total of 14 or so.  I might be slightly off.  so `1.25 ^ 14` = 22.

I'm not sure that's actually enough to be safe, since we do stand to lose the funds in the worst case.  I think a factor of 100 would be better.